addon/mixins/stateful.js

```diff
@@ -0,0 +1,299 @@
+    const actions = Object.entries(stateHash).filter(([key, val]) => isAction(key, val));
+    for (const [name, ] of actions) {
+      if (Ember.isNone(this.actions[name])) {
+        this.actions[name] = function(...args) {
+          this.send(name, ...args);
+        }
+      }
+    }
+    const subStates = findSubStates(stateHash);
+    for (const [name, hash] of subStates) {
+      this._initializeState(hash, name, fullStateName);
+    }
+  },
+
+  _initializeStateHierarchy() {
+    this._defaultStateMapping = {};
+    const actions = Object.assign({}, this.actions);
+    this._initializeState(actions, '_root', '_root');
+    this.actions._root = actions;
+  },
+
+  _defaultStateMapping: undefined,
+
+  init(...args) {
+    this._super(...args);
+    this._initializeStateHierarchy();
+    this.getStateTask('_root').perform(true);
+    const stateTaskNames = this.get('stateTaskNames') ;
+    Object.defineProperty(this, 'currentState', {
+      value: computed(...stateTaskNames.map(stateTask => `${stateTask}.isRunning`), this.getCurrentStateName),
+    });
+    // kick off the computed properties
+    this.get('state');
+    this.get('currentState');
+  },
+
+  send(actionName, ...args) {
+    let stateParts = this.get('currentState').split('.');
+    let shouldBubble = true;
+
+    while(shouldBubble && stateParts.length > 0) {
+      const action = this.get(`actions._root.${stateParts.join('.')}.${actionName}`);
+      if (action) {
+        shouldBubble = action.apply(this, args) === true;
+      }
+      stateParts.pop();
+    }
+    if (shouldBubble) {
+      const rootAction = this.actions._root[actionName];
+      if (!rootAction) {
+        throw new Error(ERRORS.NO_ROOT_ACTION(actionName, this));
+      }
+      rootAction.apply(this, args);
+    }
+  },
+
+  transitionTo(stateName) {
+    if (!this.checkIfStateExists(stateName)) {
+      throw new Error(ERRORS.NO_SUCH_STATE(stateName));
+    }
+    this.getStateTask(stateName).perform(true);
+  },
+
+  getStateTask(stateName) {
+    return this.get(getStateTaskName(stateName));
+  },
+
+  getDefaultStateName(stateName) {
+    return this._defaultStateMapping[stateName];
+  },
+
+  checkIfStateExists(stateName) {
+    const stateNames = this.get('stateNames');
+    return stateNames.includes(stateName);
+  },
+
+  *_stateTaskFunction(stateName, shouldStartDefaultSubtask = true) {
+    if (stateName !== '_root') {
+      const superStateName = getSuperStateName(stateName);
+      const superState = this.getStateTask(superStateName);
+      if (!superState.get('isRunning')) {
+        // during transitions into a deeply nested structure the superstates will not
+        // be running so kick them off first
+        superState.perform(false);
+      }
+    }
+
+    const stateActions = this.get(`actions.${stateName}`) || {};
+    try {
+      this.trigger(`try_${stateName}`);
+      yield executeStateHook(this, stateActions._try);
```

Comment:
making `executeStateHook` a generator function itself will give more flexibility (see above). We then need to use `yield* executeStateHook(this, stateActions._try);`, however...